File: android/domain/src/main/kotlin/com/bolsillo/domain/usecase/UndoLastRecord.kt
package com.bolsillo.domain.usecase

import com.bolsillo.domain.port.TransactionRepository
import javax.inject.Inject

/**
 * Reference to the most recent save so [UndoLastRecord] knows what to revert.
 * For a transfer this is the [transferGroupId]; for a single leg it's the [id].
 */
sealed class SavedRef {
    data class Single(val id: String) : SavedRef()

    data class Group(val transferGroupId: String) : SavedRef()
}

/**
 * FR 6,7 — undo the just-created record/group via soft delete (research R6).
 * Same code path as trash; the row remains restorable. Article III: no hard
 * deletes on user action.
 */
class UndoLastRecord
    @Inject
    constructor(private val transactions: TransactionRepository) {
        suspend operator fun invoke(
            lastSaved: SavedRef,
            now: Long = System.currentTimeMillis(),
        ) {
            when (lastSaved) {
                is SavedRef.Single -> transactions.softDelete(lastSaved.id, now)
                is SavedRef.Group -> transactions.softDeleteGroup(lastSaved.transferGroupId, now)
            }
        }
    }

File: ios/Packages/Bolsillo/Sources/BolsilloDomain/UseCases/UndoLastRecord.swift
import Foundation

/// Soft-deletes the most recently saved record — either a single transaction or
/// both legs of a transfer group — without hard-deleting any data
/// (Constitution Article III — no hard deletes).
///
/// The caller supplies a `SavedRef` obtained from the previous `RecordTransaction`
/// or `RecordTransfer` call. This use case is designed for the ≤5 s undo affordance
/// in the fast-recording flow (spec 001-R14).
public struct UndoLastRecord: Sendable {
    private let transactions: any TransactionRepository

    public init(transactions: some TransactionRepository) {
        self.transactions = transactions
    }

    /// Soft-delete the referenced transaction(s).
    ///
    /// - Parameters:
    ///   - lastSaved: The `SavedRef` returned by the prior record operation.
    ///   - now: The deletion timestamp (injected for testability).
    public func callAsFunction(_ lastSaved: SavedRef, now: Date) async {
        switch lastSaved {
        case .single(let id):
            await transactions.softDelete(id: id, deletedAt: now)
        case .group(let groupId):
            await transactions.softDeleteGroup(transferGroupId: groupId, deletedAt: now)
        }
    }
}
